ext/liquid_c/variable_lookup.c

```diff
@@ -0,0 +1,64 @@
+#include "liquid.h"
+#include "context.h"
+
+static ID id_has_key, id_aref, id_fetch;
+static ID id_ivar_lookups, id_ivar_name, id_ivar_command_flags;
+
+VALUE variable_lookup_evaluate(VALUE self, VALUE context)
+{
+    VALUE command_flags = 0;
+    VALUE name = rb_ivar_get(self, id_ivar_name);
+    name = context_evaluate(context, name);
+
+    VALUE object = context_find_variable(context, name, Qtrue);
+
+    VALUE lookups = rb_ivar_get(self, id_ivar_lookups);
+    Check_Type(lookups, T_ARRAY);
+
+    for (long i = 0; i < RARRAY_LEN(lookups); i++) {
+        VALUE key = context_evaluate(context, RARRAY_AREF(lookups, i));
+        VALUE next_object;
+
+        if (rb_respond_to(object, id_aref) && (
+            (rb_respond_to(object, id_has_key) && rb_funcall(object, id_has_key, 1, key)) ||
+            (rb_obj_is_kind_of(key, rb_cInteger) && rb_respond_to(object, id_fetch))
+        )) {
+            next_object = rb_funcall(object, id_aref, 1, key);
+            next_object = materialize_proc(context, object, key, next_object);
+            object = value_to_liquid_and_set_context(next_object, context);
+            continue;
+        }
+
+        if (!command_flags) {
+            command_flags = rb_ivar_get(self, id_ivar_command_flags);
+        }
+        if (NUM2LONG(command_flags) & (1 << i)) {
+            Check_Type(key, T_STRING);
+            ID intern_key = rb_intern(RSTRING_PTR(key));
```

Comment:
I don't think `rb_intern` create garbage collectable symbols.  It looks like internally [String#to_sym](https://github.com/ruby/ruby/blob/286f0b8ae3c30b5911199924a1779ca045afb079/symbol.c#L697-L732) uses `dsymbol_alloc` to allocate these strings in the `#if USE_SYMBOL_GC` conditional compilation branch and uses `intern_str` (which `rb_intern` uses) in the `#else` branch.

If you want to create a dynamic symbol, then it looks like `rb_intern_str` is the C API for doing that.

It looks like we won't be able to use C APIs that take an `ID`, since `rb_sym2id` says it [will make it a permanent object](https://github.com/ruby/ruby/blob/286f0b8ae3c30b5911199924a1779ca045afb079/symbol.c#L749), so I don't think we will be able to use `rb_respond_to` either.

Reply:
The possible commands that could take this path are defined in https://github.com/Shopify/liquid/blob/b0629f17f79a96eb7001b55ea0877b14b53b67a0/lib/liquid/variable_lookup.rb#L4, so the set is finite and small here.

If we make a new struct for the C variable lookup implementation, we could keep track of the ID at parse time in https://github.com/Shopify/liquid-c/blob/69a4715cbbc965b6c1768dcdbcda22f1784ec7ae/ext/liquid_c/parser.c#L113-L114

then we wouldn't need any dynamic creation of symbols or IDs for this. Do you think it's worth it over the current code, which incurs the cost of a `st_lookup` in the symbol table?

Reply:
Actually the best option is probably to factor those `rstring_eq` calls into a new function that returns the ID directly, and use it in both places.

Reply:
> The possible commands that could take this path are defined in https://github.com/Shopify/liquid/blob/b0629f17f79a96eb7001b55ea0877b14b53b67a0/lib/liquid/variable_lookup.rb#L4, so the set is finite and small here.

Oops, I forgot about that.  Nevermind then.

We could keep it like this for now and optimize this better once we replace this class more completely with a C implementation.